vnext/Microsoft.ReactNative/CoalescingEventEmitter.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CoalescingEventEmitter.h"
+#include "CoalescingEventEmitter.g.cpp"
+#include "DynamicWriter.h"
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+/*static*/ ReactPropertyId<CoalescingEventEmitter> CoalescingEventEmitter::Property() noexcept {
+  static ReactPropertyId<CoalescingEventEmitter> propId{L"ReactNative.CoalescingEventEmitter", L"Emitter"};
+  return propId;
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    Mso::React::IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+CoalescingEventEmitter::CoalescingEventEmitter(Mso::CntPtr<Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {}
+
+void CoalescingEventEmitter::EmitJSEvent(int64_t coalesceKey, JSValueArgWriter const &paramsArgWriter) noexcept {
+  std::unique_lock lock(m_mutex);
+
+  for (auto &evt : m_eventQueue) {
+    if (evt.coalesceKey && *evt.coalesceKey == coalesceKey) {
+      evt.expired = true;
+    }
+  }
+
+  auto paramsWriter = winrt::make_self<DynamicWriter>();
+  paramsArgWriter(*paramsWriter);
+
+  CoalescingEvent newEvent;
+  newEvent.coalesceKey = coalesceKey;
+  newEvent.args = paramsWriter->TakeValue();
+
+  m_eventQueue.push_back(std::move(newEvent));
+  EnsureRenderCallbackRegistered();
+}
+
+void CoalescingEventEmitter::EmitDurableJSEvent(JSValueArgWriter const &paramsArgWriter) noexcept {
+  std::unique_lock lock(m_mutex);
+
+  auto paramsWriter = winrt::make_self<DynamicWriter>();
+  paramsArgWriter(*paramsWriter);
+
+  CoalescingEvent newEvent;
+  newEvent.args = paramsWriter->TakeValue();
+
+  m_eventQueue.push_back(std::move(newEvent));
+  EnsureRenderCallbackRegistered();
+}
+
+void CoalescingEventEmitter::EnsureRenderCallbackRegistered() noexcept {
+  if (!m_renderingRevoker) {
+    m_renderingRevoker = xaml::Media::CompositionTarget::Rendering(
+        winrt::auto_revoke, [weakThis{get_weak()}](auto const &, auto const &) {
+          if (auto strongThis = weakThis.get()) {
+            strongThis->OnRendering();
+          }
+        });
+  }
+}
+
+void CoalescingEventEmitter::OnRendering() noexcept {
+  std::unique_lock lock(m_mutex);
+
+  // Submit the batch of requests to the JS queue
+  while (!m_eventQueue.empty()) {
+    auto &evt = m_eventQueue.front();
+    if (!evt.expired) {
+      m_context->CallJSFunction("RCTEventEmitter", "receiveEvent", std::move(evt.args));
```

Comment:
`OnRendering` is happening in the ui thread; can we avoid locking on the mutex? (instead try to acquire it, if it fails leave it to the next frame to try again, for example)

Reply:
The locked section adding to queue should be near instant, so I don't expect contention to be likely. If there is contention, the function should finish quickly and let the UI thread resume. I can scope it more though.

Delaying by a frame would have the potential to add more noticeable delays.

Reply:
Actually this whole thing should be dispatching to JS